Code/GraphMol/MolOps.cpp

```diff
@@ -683,15 +680,99 @@ std::vector<ROMOL_SPTR> getMolFrags(const ROMol &mol, bool sanitizeFrags,
           atomsInFrag.set(idx);
         }
       }
-      for (unsigned int idx = 0; idx < mol.getNumAtoms(); ++idx) {
-        if (!atomsInFrag[idx]) {
-          frag->removeAtom(idx);
+      auto fragmentHasChallengingFeatures =
+          [&](const INT_VECT &comp,
+              const boost::dynamic_bitset<> &atomsInFrag) -> bool {
+        for (auto idx : comp) {
+          // check for atoms with stereochem:
+          const auto atom = mol.getAtomWithIdx(idx);
+          if (atom->getChiralTag() != Atom::ChiralType::CHI_UNSPECIFIED &&
+              atom->getChiralTag() != Atom::ChiralType::CHI_OTHER) {
+            return true;
+          }
+          for (auto bnd : mol.atomBonds(atom)) {
+            if (atomsInFrag[bnd->getOtherAtomIdx(idx)]) {
+              if (bnd->getStereo() != Bond::BondStereo::STEREONONE &&
+                  bnd->getStereo() != Bond::BondStereo::STEREOANY) {
+                return true;
+              }
+            }
+          }
+        }
+        for (auto sgroup : getSubstanceGroups(mol)) {
+          for (auto aid : sgroup.getAtoms()) {
+            if (atomsInFrag[aid]) {
+              return true;
+            }
+          }
+          for (auto aid : sgroup.getParentAtoms()) {
+            if (atomsInFrag[aid]) {
+              return true;
+            }
+          }
+        }
+        // doesn't seem like this should be necessary, but in case
+        // we ever need stereogroups where the atoms aren't marked
+        // with stereo...
+        for (auto stereoGroup : mol.getStereoGroups()) {
+          for (auto atom : stereoGroup.getAtoms()) {
+            if (atomsInFrag[atom->getIdx()]) {
+              return true;
+            }
+          }
+        }
+        return false;
+      };
+      if (comp.size() == 1 ||
+          (nFrags > 3 && !fragmentHasChallengingFeatures(comp, atomsInFrag))) {
```

Comment:
So, if we have a protein with, lets say, a DAT SGroup containing some label (which is something I could see happening), we'll still hit the pathologically slow case.

Seeing that general case may be pathologically slow, and that we need to write special treatment for specific cases to avoid it, I wonder if the change in #7056 was really worth it.

Reply:
I think the issue is that removeAtomFromGroups is an N**2 op called for every atom removal.  I think I could make it work with batch edit operations that would be a lot more efficient.

Should I give it a go?

Reply:
> So, if we have a protein with, lets say, a DAT SGroup containing some label (which is something I could see happening), we'll still hit the pathologically slow case.

If there's a protein with DAT SGroups on a bunch of smaller fragments (i.e. waters) you'll hit the super slow case. If it's the protein itself, or the ligand, everything should be fine. I'll add a test for that to demonstrate and to be sure

> Seeing that general case may be pathologically slow, and that we need to write special treatment for specific cases to avoid it, I wonder if the change in #7056 was really worth it.

The primary reason for the change in #7056 was to allow S groups to be preserved in `getMolFrags()` when some of their atoms are removed. The code to do that was quite complex, this combined with the the complexity already present in `getMolFrags()`  seemed like too much to me. I think the new version is faster in the most common case, small molecules that don't have tens or hundreds of fragments, and it is still simpler.

Reply:
> I think the issue is that removeAtomFromGroups is an N**2 op called for every atom removal. 

I don't believe that this is actually the slow part.

Reply:
I'll run a profile on this MR to verify.

Reply:
Interestingly, the removal of computed properties is high on the profile, and removeAtomsFromGroups is nowhere to be seen.

<img width="848" alt="image" src="https://github.com/rdkit/rdkit/assets/9888416/96b11720-bb15-450c-8198-a3ff431ae992">

Reply:
This might be an avenue for a future optimization.  I was surprised to see it this high (might be a build for profile issue though)

Reply:
@greglandrum and I will address this in a separate MR.